File: 239.sliding-window-maximum.py
#
# @lc app=leetcode id=239 lang=python3
#
# [239] Sliding Window Maximum
#

# @lc code=start
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        h = []
        cnt = {}
        for i in range(k-1):
            heappush(h, -nums[i])
            if nums[i] in cnt:
                cnt[nums[i]] += 1
            else:
                cnt[nums[i]] = 1

        res = []
        for i in range(k-1, len(nums)):
            # add new element
            heappush(h, -nums[i])
            if nums[i] in cnt:
                cnt[nums[i]] += 1
            else:
                cnt[nums[i]] = 1

            # pop inexist top element in heap
            while -h[0] not in cnt or cnt[-h[0]] == 0:
                heappop(h)
                
            res.append(-h[0])

            # pop front
            cnt[nums[i-k+1]] -= 1

        return res
# @lc code=end

File: 239.sliding-window-maximum.cpp
/*
 * @lc app=leetcode id=239 lang=cpp
 *
 * [239] Sliding Window Maximum
 */

// @lc code=start
class Solution {
public:
    // O(nlogn)
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res(nums.size() - k + 1, 0);
        priority_queue<int> pq;
        map<int, int> cnt;
        for(int i=0;i<k-1;i++){
            pq.push(nums[i]);
            cnt[nums[i]] = cnt[nums[i]] + 1;
        }
        for(int i=k-1;i<nums.size();i++){
            // insert new element
            pq.push(nums[i]);
            cnt[nums[i]] = cnt[nums[i]] + 1;

            // get maximum
            while(!cnt[pq.top()]){
                pq.pop();
            }
            res[i-k+1] = pq.top();

            // pop front
            cnt[nums[i-k+1]] = cnt[nums[i-k+1]] - 1;;
        }
        return res;
    }
};
// @lc code=end

